gwcs/wcs.py

```diff
@@ -634,3 +641,273 @@ def fix_inputs(self, fixed):
+        inv_degree : int
+            Degree of the inverse SIP polynomial. If supplied max_inv_pixel_error
+            is ignored.
+        verbose : bool
+            print progress of fits
+
+        Returns
+        -------
+        FITS header with all SIP WCS keywords
+
+        Raises
+        ------
+        ValueError
+            If the WCS is not 2D, an exception will be raised. If the specified accuracy
+            (both forward and inverse, both rms and maximum) is not achieved an exception
+            will be raised.
+
+        Notes
+        -----
+
+        Use of this requires a judicious choice of required accuracies. Attempts to use
+        higher degrees (~7 or higher) will typically fail due floating point problems
+        that arise with high powers.
+
+
+        """
+        if (self.forward_transform.n_inputs != 2
+                or self.forward_transform.n_outputs != 2):
+            raise ValueError(
+                "The to_fits_sip method only works with 2-dimensional transforms")
+
+        transform = self.forward_transform
+        # Determine reference points.
+        (xmin, xmax), (ymin, ymax) = bounding_box
+        crpix1 = int((xmax - xmin) / 2)
+        crpix2 = int((ymax - ymin) / 2)
+        crval1, crval2 = transform(crpix1, crpix2)
+        hdr = fits.Header()
+        hdr['naxis'] = 2
+        hdr['naxis1'] = xmax
+        hdr['naxis2'] = ymax
+        hdr['ctype1'] = 'RA---TAN-SIP'
+        hdr['ctype2'] = 'DEC--TAN-SIP'
+        hdr['CRPIX1'] = crpix1 + 1
+        hdr['CRPIX2'] = crpix2 + 1
+        hdr['CRVAL1'] = crval1
+        hdr['CRVAL2'] = crval2
+        hdr['cd1_1'] = 0  # Placeholders
+        hdr['cd1_2'] = 0
+        hdr['cd2_1'] = 0
+        hdr['cd2_2'] = 0
+        # Now rotate to native system and deproject. Recall that transform
+        # expects pixels in the original coordinate system, but the SIP
+        # transform is relative to crpix coordinates, thus the initial shift.
+        ntransform = ((Shift(crpix1) & Shift(crpix2)) | transform
+                      | RotateCelestial2Native(crval1, crval2, 180)
+                      | Sky2Pix_TAN())
+        u, v = _make_sampling_grid(npoints, bounding_box)
+        undist_x, undist_y = ntransform(u, v)
+        # double sampling to check if sampling is sufficient
+        ud, vd = _make_sampling_grid(2 * npoints, bounding_box)
+        undist_xd, undist_yd = ntransform(ud, vd)
+        # undist_x, undist_y = ntransform(x, y)
+        # Determine approximate pixel scale in order to compute error threshold
+        # from the specified pixel error. Computed at the center of the array.
+        x0, y0 = ntransform(0, 0)
+        xx, xy = ntransform(1, 0)
+        yx, yy = ntransform(0, 1)
+        dist1 = np.sqrt((xx - x0)**2 + (xy - y0)**2)
+        dist2 = np.sqrt((yx - x0)**2 + (yy - y0)**2)
+        # Average the distances in the two different directions.
+        plate_scale = (dist1 + dist2) / 2.
```

Comment:
Personally, I am not a fan of arithmetic average for this _specific_ measurement. I think `sqrt(jacobian)` is a better estimate of the pixel scale (assumes _square_ pixels)

Reply:
By the way, this pixel scale is computed `1/2` pixel off from the center of the array.

A suggested approach of using `sqrt(pixel_area)` could be implemented following https://github.com/spacetelescope/tweakwcs/blob/9cee6794ce78e1155b473700a87f10681f2f83ab/tweakwcs/tpwcs.py#L210-L242 example with `self.det_to_tanp` replaced with ntransform(xs + crpix1, ys + crpix2)`.

Reply:
I suggest we get this right into astropy so everyone can use it. I'll wait and include it when it is available. I don't wish to cut and paste it resulting in code duplication.

Reply:
It can't be in `astropy` because APE 14 does not deal with the "tangent" plane. But if you want this for FITS WCS, see https://docs.astropy.org/en/stable/api/astropy.wcs.utils.proj_plane_pixel_area.html#astropy.wcs.utils.proj_plane_pixel_area. However, this is the area *after* the CD matrix has been applied. (CD itself may contain SIP's linear terms so it is difficult to disentangle these).

I think this is quite a specialized think and you may not have a choice but to duplicate that code (assuming you want to adopt this approach of estimating pixel scale).

Reply:
A simpler approach but closer is np.sqrt(delta_pix_x * delta_pix_y). Would you be happy with that?

Reply:
It is not about my happiness. I was just suggesting how to improve the estimate (if it is important). If not - anything works. What you are proposing is the upper estimate of the pixel scale (it assumes there is no "skew", or that the pixels are *rectangular*).

Reply:
True, but skew's effect on area is a cosine effect. You would really have to have a lot of it to make a significant difference. More likely one would be affected more by differences in plate scale in the two different directions than this.

Reply:
I agree. It is up to you to decide if the accuracy of the computed scale is critical or not.

Reply:
It's not. But I did compute the cross product area and use that. I want you happy :-)  But for the half pixel, that's where I draw the line. If that matters, the user has much bigger problems. Updated PR will be generated.